### ISSUE TITLE
3.7.2 - traceback on extracting job with unicode ? in filename

### ISSUE THREAD

**Issue by thezoggy** [sabnzbd/sabnzbd#2462]
discord user reported issue with sab 3.7.2 not handling a download on windows:
```
2023-02-17 22:06:07,992::DEBUG::[assembler:85] Decoding part of \\?\D:\Cache\In-Progress\[GS] Netoge no Yome wa Onnanoko ja Nai to Omotta! 02 (BD 1080p 10bit AAC) [7ACD4B5A]\[GS] Netoge no Yome wa Onnanoko ja Nai to Omotta 02 (BD 1080p 10bit AAC) [7ACD4B5A].mkv
2023-02-17 22:06:07,993::INFO::[notifier:123] Sending notification: Error - Disk error on creating file D:\Cache\In-Progress\[GS] Netoge no Yome wa Onnanoko ja Nai to Omotta! 02 (BD 1080p 10bit AAC) [7ACD4B5A]\[GS] Netoge no Yome wa Onnanoko ja Nai to Omotta 02 (BD 1080p 10bit AAC) [7ACD4B5A].mkv (type=error, job_cat=None)
2023-02-17 22:06:07,993::ERROR::[assembler:113] Disk error on creating file D:\Cache\In-Progress\[GS] Netoge no Yome wa Onnanoko ja Nai to Omotta! 02 (BD 1080p 10bit AAC) [7ACD4B5A]\[GS] Netoge no Yome wa Onnanoko ja Nai to Omotta 02 (BD 1080p 10bit AAC) [7ACD4B5A].mkv
2023-02-17 22:06:07,994::INFO::[assembler:116] Winerror: None - 0xc0000033
2023-02-17 22:06:07,994::INFO::[assembler:121] Traceback: 
Traceback (most recent call last):
  File "sabnzbd\assembler.py", line 86, in run
  File "sabnzbd\assembler.py", line 178, in assemble
OSError: [Errno 22] Invalid argument: '\\\\?\\D:\\Cache\\In-Progress\\[GS] Netoge no Yome wa Onnanoko ja Nai to Omotta! 02 (BD 1080p 10bit AAC) [7ACD4B5A]\\[GS] Netoge no Yome wa Onnanoko ja Nai to Omotta\x1f 02 (BD 1080p 10bit AAC) [7ACD4B5A].mkv'
2023-02-17 22:06:07,994::INFO::[downloader:395] Pausing
```

looking it looks to be the unicode character in the yenc/filename that causes it to puke. assemble dies at the start for it.

we of course filter out illegal stuff by default on windows.. but looks like we might need to add this one to the list
https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/filesystem.py#L203

sending nzb to ya via email

**Comment by thezoggy** [comment]
trying nzb out with newsbin, it extracts and replaces the unicode ! with _ in the final name
![newsbin](http://zoggy.net/explorer_2023-02-17_17-16-06.png)

**Comment by thezoggy** [comment]
while the nzb completes and works just fine in newsbin, trying it in sab develop latest it says articles arent there and aborts the job... which i know is incorrect because i can use something other than sab and load the nzb and it works fine.

```
2023-02-17 17:25:59,885::INFO::[nzbstuff:232] Article JzAkLsZxPoRkCnZsUeTpOxOy-1483555129956@nyuu unavailable on all servers, discarding
2023-02-17 17:25:59,885::DEBUG::[nzbstuff:1510] Availability ratio=105.05, bad articles=0, total bytes=1560736426, missing bytes=736544, par2 bytes=75796569
2023-02-17 17:25:59,886::INFO::[nzbstuff:232] Article JpPfYuHkEaPsPxMsCcDePkBz-1483555394849@nyuu unavailable on all servers, discarding
2023-02-17 17:25:59,886::DEBUG::[nzbstuff:1510] Availability ratio=105.00, bad articles=1, total bytes=1560736426, missing bytes=1476800, par2 bytes=75796569
2023-02-17 17:25:59,886::INFO::[nzbstuff:232] Article ZbKyVfGmCzOwEuIpVtGtSbOa-1483555390878@nyuu unavailable on all servers, discarding
2023-02-17 17:25:59,887::DEBUG::[nzbstuff:1510] Availability ratio=104.96, bad articles=2, total bytes=1560736426, missing bytes=2217119, par2 bytes=75796569
...
2023-02-17 17:26:05,068::INFO::[nzbqueue:791] [sabnzbd.nzbqueue.register_article] Ending job GS.Netoge.no.Yome.wa.Onnanoko.ja.Nai.to.Omotta.02.BD.1080p.10bit.AAC.7ACD4B5A
2023-02-17 17:26:05,068::INFO::[nzbstuff:232] Article CwNvUuEyXkIhQxMxGaDzIkDa-1483555141072@nyuu unavailable on all servers, discarding
2023-02-17 17:26:05,068::INFO::[nzbqueue:392] [sabnzbd.assembler.run] Removing job GS.Netoge.no.Yome.wa.Onnanoko.ja.Nai.to.Omotta.02.BD.1080p.10bit.AAC.7ACD4B5A
2023-02-17 17:26:05,068::DEBUG::[nzbstuff:1510] Availability ratio=100.12, bad articles=100, total bytes=1560736426, missing bytes=74044485, par2 bytes=75796569
2023-02-17 17:26:05,069::INFO::[nzbqueue:233] Saving queue
2023-02-17 17:26:05,069::DEBUG::[nzbstuff:1221] Abort job "GS.Netoge.no.Yome.wa.Onnanoko.ja.Nai.to.Omotta.02.BD.1080p.10bit.AAC.7ACD4B5A", due to impossibility to complete it
```

**Comment by puzzledsab** [comment]
Seems like it has ascii value 31 (hex 1f) which is not valid. Everything below 32 should be converted.

**Comment by thezoggy** [comment]
@puzzledsab if you jump on sab discord i can share nzb with ya, or if you want you can pull it down from geek if you have that indexer.

**Comment by thezoggy** [comment]
interestingly if i grab the nzb from geek myself for ep 01.. everything good in the end with latest develop branch

release:
`[GS] Netoge no Yome wa Onnanoko ja Nai to Omotta? 01 (BD 1080p 10bit AAC) [E74233AF]`

when grabbing the nzb from the site, the `？` is not there in the .nzb
`[GS].Netoge.no.Yome.wa.Onnanoko.ja.Nai.to.Omotta .01.(BD.1080p.10bit.AAC).[E74233AF].nzb`

i assume that the nzb provided by the user prob came from sonarr or something.. but not sure how the .nzb would be the issue.

I see sab dev process it just fine, end result does include the unicode ? and all is good:
![sab-dev](http://zoggy.net/explorer_2023-02-17_20-36-58.png)

**Comment by sanderjo** [comment]
No problem with that "7ACD4B5A" NZB on Linux.
I'll create a post with 0x1f in the resulting filename, so it's easier to test

**Comment by thezoggy** [comment]
trying out sander's nzb which is a little different. original issue the file is not in a rarset.

with sanders rarset, its got a different outcome as the assembler is fine with the rarset but then the stuff breaks directunpacker and the job fails with the "Unpacking failed, write error or disk is full" error.
```
2023-02-18 01:57:54,447::INFO::[directunpacker:220] Error in DirectUnpack of weird_char: Cannot create \\?\C:\sab\downloads\complete\raw\_UNPACK_reftestnzb.weird.0x1f.in.filename.4fdcb05d2b4f\Hello Some Ugly  Char 2023.bin
2023-02-18 01:57:54,447::INFO::[directunpacker:454] Aborting DirectUnpack for weird_char
2023-02-18 01:57:54,863::DEBUG::[filesystem:941] Removing dir recursively \\?\C:\sab\downloads\complete\raw\_UNPACK_reftestnzb.weird.0x1f.in.filename.4fdcb05d2b4f
2023-02-18 01:57:54,865::DEBUG::[directunpacker:334] DirectUnpack Unrar output 
UNRAR 6.20 x64 freeware      Copyright (c) 1993-2023 Alexander Roshal


Extracting from \\?\C:\sab\downloads\incomplete\reftestnzb_weird_0x1f_in_filename_4fdcb05d2b4f\weird_char.part1.rar

Extracting  \\?\C:\sab\downloads\complete\raw\_UNPACK_reftestnzb.weird.0x1f.in.filename.4fdcb05d2b4f\All OK Nothing Wrong.bin  OK
Cannot create \\?\C:\sab\downloads\complete\raw\_UNPACK_reftestnzb.weird.0x1f.in.filename.4fdcb05d2b4f\Hello Some Ugly  Char 2023.bin
2023-02-18 01:57:55,667::INFO::[downloader:437] Waiting for post-processing to finish
```
full log: https://gist.github.com/thezoggy/3fa48538f2a18fd5ed52d2debaf092c0

**Comment by thezoggy** [comment]
@puzzledsab thanks for the fix! can confirmed it resolved original user issue. resulting filename has `_` like other apps. 
previously the ep01 did download just fine and the resulting filename had the unicode ?, now post this update it now has the character converted to a `_`. So a change from before, but atleast everything looks to be handled consistently. Which probably is better in the long run.

Now for sanders test nzb, that still fails due to the whole "Cannot create" being seen in the unrar output. Which looks like has been something in the code for 6+ years. IMHO we should just remove this string because its a warning and as I showed above unrar these days just replaces the character itself with a `_` and we could have just continued just fine.

Now do we just drop the "Cannot create" entry or try to add some logic to keep it but only really abort if "WARNING: Attempting to correct the invalid file or directory name" isnt seen?
https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/directunpacker.py#L206

**Comment by Safihre** [comment]
It seems I should try to fix #1574 after all, it's the same issue.
